src/components/chart/down-sample.ts

```diff
@@ -0,0 +1,67 @@
+import type { LineSeriesOption } from "echarts";
+
+export function downSampleLineData(
+  data: LineSeriesOption["data"],
+  chartWidth: number,
+  minX?: number,
+  maxX?: number
+) {
+  if (!data || data.length < 10) {
+    return data;
+  }
+  const width = chartWidth * window.devicePixelRatio;
+  if (data.length <= width) {
+    return data;
+  }
+  const min = minX ?? data[0]![0];
+  const max = maxX ?? data[data.length - 1]![0];
+  const step = Math.floor((max - min) / width);
+  const frames = new Map<
+    number,
+    {
+      min: { point: (typeof data)[number]; x: number; y: number };
+      max: { point: (typeof data)[number]; x: number; y: number };
+    }
+  >();
+
+  // Group points into frames
+  for (const point of data) {
+    const pointData =
+      point && typeof point === "object" && "value" in point
```

Comment:
Here we assume point can be an object, but on line 16 and 17 we assume it is an array?

Reply:
Fixed. I wonder if we should even bother with this case. We don't use or need object values for line charts and they slow down processing. But echarts supports this and typing would be complicated if we disallow it.